lib/Dialect/Triton/Transforms/LoopUnroll.cpp

```diff
@@ -22,24 +23,65 @@
 
 namespace mlir::triton {
 
-static const char *loopUnrollFactorAttrName = "tt.loop_unroll_factor";
-
 namespace {
 
 class LoopUnrollPass : public TritonLoopUnrollBase<LoopUnrollPass> {
 
   int getUnrollFactorOrDefault(scf::ForOp forOp) {
     // Use the attribute attached to the loop if it exists otherwise set the
     // factor to 1 to suppress the unrolling.
-    if (auto factor = forOp->getAttrOfType<IntegerAttr>(
-            mlir::triton::loopUnrollFactorAttrName))
+    if (auto factor =
+            forOp->getAttrOfType<IntegerAttr>(loopUnrollFactorAttrName))
       return factor.getInt();
     return 1;
   }
 
+  int getUnrollIdOrDefault(scf::ForOp forOp) {
+    // Use the attribute attached to the loop if it exists otherwise set the
+    // factor to 1 to suppress the unrolling.
+    if (auto factor = forOp->getAttrOfType<IntegerAttr>(unrolledLoopIdAttrName))
+      return factor.getInt();
+    return 0;
+  }
+
+  const char *loopUnrollFactorAttrName = "tt.loop_unroll_factor";
+  const char *unrolledLoopIdAttrName = "tt.unrolled_loop_id";
+  const char *pipelineStagesAttrName = "tt.num_stages";
+
 public:
   LoopUnrollPass() = default;
   LoopUnrollPass(const LoopUnrollPass &) {}
+
+  SmallVector<scf::ForOp, 2> getUnrolledLoopsAndClearAttrs(unsigned loopId) {
+    SmallVector<scf::ForOp, 2> loops;
+    getOperation()->walk([&](scf::ForOp forOp) {
+      if (getUnrollIdOrDefault(forOp) == loopId)
+        loops.push_back(forOp);
+    });
+
+    assert(loops.size() <= 2 && "Expect at most 2 loops, one for the main loop "
+                                "and one for the prolog/epilog");
+    SmallVector<int, 2> loopInstructionCounts;
+    for (auto loop : loops) {
+      loop->removeAttr(loopUnrollFactorAttrName);
+      loop->removeAttr(unrolledLoopIdAttrName);
+      int count = 0;
+      loop->walk([&](Operation *op) { count++; });
+      loopInstructionCounts.push_back(count);
+    }
+    if (loops.size() == 2) {
+      // check which one is the unrolled loop and which one is the prolog/epilog
+      // loop. A simple heuristic is to check the number of instructions in the
+      // loop. The unrolled main loop should have the most instructions.
+      // sort the loops by the number of instructions. The unrolled main loop
+      // should go first.
+      if (loopInstructionCounts[0] < loopInstructionCounts[1])
+        std::swap(loops[0], loops[1]);
+    }
+
+    return loops;
+  }
```

Comment:
this seems pretty convoluted. Can we improve `loopUnrollByFactor` instead to either return the loops or improve the existing annotator callback to handle the loop operations.

Reply:
> Good point. We should be able to extend the parameter list of loopUnrollByFactor to take a by-ref argument that'll be populated with the resulting loops. Changing the return value might be a breaking change. WDYT?

yes I think it is reasonable change to return the new loops even if it slightly changes the interface

Reply:
Sounds good. https://github.com/llvm/llvm-project/pull/114573 is to extend the return type on the mlir side.